synapse/handlers/receipts.py

```diff
@@ -147,23 +147,63 @@ async def received_client_receipt(
             receipt_type=receipt_type,
             user_id=user_id,
             event_ids=[event_id],
-            data={"ts": int(self.clock.time_msec())},
+            data={"ts": int(self.clock.time_msec()), "hidden": hidden},
         )
 
         is_new = await self._handle_new_receipts([receipt])
         if not is_new:
             return
 
-        if self.federation_sender:
+        if self.federation_sender and not hidden:
             await self.federation_sender.send_read_receipt(receipt)
 
 
 class ReceiptEventSource:
     def __init__(self, hs: "HomeServer"):
         self.store = hs.get_datastore()
 
+    def _filter_out_hidden(
+        self, events: List[JsonDict], user_id: str
+    ) -> List[JsonDict]:
+        visible_events = []
+
+        # filter out hidden receipts the user shouldn't see
+        for event in events:
+            content = event.get("content", {})
+            new_event = event.copy()
+            new_event["content"] = {}
+
+            for event_id in content.keys():
+                event_content = content.get(event_id, {})
+                m_read = event_content.get("m.read", {})
+
+                # If m_read is missing copy over the original event_content as there is nothing to process here
+                if not m_read:
+                    new_event["content"][event_id] = event_content
+                    continue
+
+                new_users = {}
+                for rr_user_id in m_read.keys():
+                    user_rr = m_read[rr_user_id]
+                    hidden = user_rr.get("hidden", False)
+                    if not hidden or rr_user_id == user_id:
+                        new_users[rr_user_id] = user_rr
+                        if hidden:
+                            new_users[rr_user_id].pop("hidden")
+                            new_users[rr_user_id]["org.matrix.msc2285.hidden"] = True
```

Comment:
This isn't exactly nice. Is preferable to use a prefixed key in the RR data or should we be doing this?